source/inputoutput.rst

```diff
@@ -3,15 +3,113 @@ Inputs and Outputs in BharatSim
 
 Inputs
 ------
-**to be completed**
 
-* function should feed in the csv file 1 row at a time
-* data gets 'ingested' and graphNodes are created
+BharatSim uses a `CSV <https://en.wikipedia.org/wiki/Comma-separated_values>`_ file as an input. It is equipped to **ingest** data from a file, by reading it and converting the data to the network.
+
+In order to ingest data, we need to make use of three functions:
+
+* ``ingestData``, which is a method of the ``simulation`` class
+* ``ingestCSVData``, a method of the ``ContextBuilder`` object
+* A user-defined function which tells the program what data to extract and what to do with it
+
+Using the framework-defined functions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+First, we need to import the necessary packages:
+
+.. code-block:: scala
+
+  import com.bharatsim.engine.ContextBuilder._
+  import com.bharatsim.engine.execution.Simulation
+  import com.bharatsim.engine.graph.ingestion.{GraphData, Relation}
+
+.. hint:: You'll see why we import ``GraphData`` and ``Relation`` in the next section!
+
+The next step is to create an instance of the simulation class in the main function,
+
+.. code-block:: scala
+
+  val simulation = Simulation()
+
+We then ingest the data in the following way:
+
+.. code-block:: scala
+
+  simulation.ingestData(implicit context => {
+    ingestCSVData("citizen10k.csv", csvDataExtractor)
+    logger.debug("Ingestion done")
+  })
+
+where ``csvDataExtractor`` is the user-defined function.
```

Comment:
Call this user-defined function something like ``myCsvDataExtractor`` or something, just like you did for the ``myCsvSpecs`` function in the output section. One of the big problems that beginners have is distinguishing between framework and modeller-defined functions.